### PR TITLE
[SIW-1334] Implemented apim resources

### PR DESCRIPTION
**Pull request by christian-calabrese**
#### List of Changes
Implemented APIM resources to expose user function

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: infra/resources/_modules/apim/api/ioweb/user-function/_swagger.json [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    "version": "1.0.0"
+  },
+  "paths": {
+    "/wallet-instances/current/status": {
+      "get": {
+        "summary": "Retrieve the current Wallet Instance status",
+        "operationId": "getCurrentWalletInstanceStatus",
+        "security": [
+          {
+            "FunctionsKey": [],
```
Comment (silvicir) [inline infra/resources/_modules/apim/api/ioweb/user-function/_swagger.json:14]: should we remove `FunctionsKey`?

Comment (mamu0) [reply infra/resources/_modules/apim/api/ioweb/user-function/_swagger.json:14]: Actually we have used the same swagger configuration of the user_func, but if is not necessary we can remove the entire `security` block for both

Comment (silvicir) [reply infra/resources/_modules/apim/api/ioweb/user-function/_swagger.json:14]: but unlike the function, in the APIM it's not necessary to use the` x-functions-key` header for authentication, right?

File: infra/resources/_modules/apim/apim.tf [hunk 2]
```diff
@@ -0,0 +1,37 @@
+module "apim_v2_wallet_api" {
+  source = "git::https://github.com/pagopa/terraform-azurerm-v3//api_management_api?ref=v8.12.2"
+
+  name                  = format("%s-ioweb-wallet", var.project_legacy)
```
Comment (Krusty93) [inline infra/resources/_modules/apim/apim.tf:4]:
```suggestion
  name                  = "IO Web - Wallet"
```

File: infra/resources/prod/main.tf [hunk 3]
```diff
@@ -132,3 +132,23 @@ module "iam" {
 
   cdn_storage_account_id = module.cdn.storage_account_cdn.id
 }
+
+module "apim" {
+  source = "../_modules/apim"
+
+  project_legacy = local.project_legacy
+  apim = {
+    name                = local.apim.name
```
Comment (Krusty93) [inline infra/resources/prod/main.tf:141]: Can we use a `data` to ensure we have the right references?

File: infra/resources/_modules/apim/apim.tf [hunk 4]
```diff
@@ -0,0 +1,37 @@
+module "apim_v2_wallet_api" {
```
Comment (Krusty93) [inline infra/resources/_modules/apim/apim.tf:1]:
```suggestion
module "apim_v2_web_wallet_api" {
```

File: infra/resources/_modules/apim/apim.tf [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  source = "git::https://github.com/pagopa/terraform-azurerm-v3//api_management_api?ref=v8.12.2"
+
+  name                  = format("%s-ioweb-wallet", var.project_legacy)
+  api_management_name   = var.apim.name
+  resource_group_name   = var.apim.resource_group_name
+  product_ids           = [var.product_id]
+  subscription_required = false
```
Comment (Krusty93) [inline infra/resources/_modules/apim/apim.tf:8]: Should these APIs be accessible from external world?

File: infra/resources/_modules/apim/apim.tf [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+  service_url = format("https://%s", var.user_function.function_hostname)
+
+  description  = "Wallet APIs"
+  display_name = "IO Web - Wallet"
+  path         = "ioweb/wallet/api/v1"
```
Comment (Krusty93) [inline infra/resources/_modules/apim/apim.tf:14]: I suggest to remove `v1` from the base path, but instead use `revision` variable to create a revision with name `v1`, in order to better support API versioning

File: infra/resources/_modules/apim/apim.tf [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+  xml_content = file("${path.module}/api/ioweb/user-function/_base_policy.xml")
+}
+
+resource "azurerm_api_management_named_value" "user_func_key" {
+  name                = "io-wallet-user-func-key"
+  api_management_name = var.apim.name
+  resource_group_name = var.apim.resource_group_name
+  display_name        = "io-wallet-user-func-key"
+  value               = data.azurerm_key_vault_secret.funciowallet_default.value
```
Comment (Krusty93) [inline infra/resources/_modules/apim/apim.tf:30]: Can we get the key from function module outputs?

File: infra/resources/_modules/apim/api/ioweb/user-function/_swagger.json (hunk 1 above)
Comment (Krusty93) [reply infra/resources/_modules/apim/api/ioweb/user-function/_swagger.json:14]: @silvicir it depends on the `subscription_required` flag set at API group level. In this PR, it is disabled and then there is no authentication

File: infra/resources/prod/main.tf (hunk 3 above)
Comment (mamu0) [reply infra/resources/prod/main.tf:141]: The `name` and `resource group` are essential values for constructing the data for `APIM`. Given that these values are already required for this purpose, define a data to retrieve the same two outputs might be redundant and unnecessary.

File: infra/resources/_modules/apim/apim.tf (hunk 7 above)
Comment (mamu0) [reply infra/resources/_modules/apim/apim.tf:30]: Currently, in the [function_app](https://github.com/pagopa/dx/blob/main/infra/modules/azure_function_app/outputs.tf) module being used, it is not among the outputs.

File: infra/resources/_modules/apim/apim.tf (hunk 2 above)
Comment (mamu0) [reply infra/resources/_modules/apim/apim.tf:4]: i've seen that in other configurations (E.g. io-infra) is used the `format("%s-NAME", product)` form, maybe is better to maintain the seme format.

File: infra/resources/_modules/apim/apim.tf (hunk 5 above)
Comment (silvicir) [reply infra/resources/_modules/apim/apim.tf:8]: no just from IO web

**Review by silvicir — APPROVED** [review]

**Pull request merged**